강원대 BE 손상욱 - Step2 회원 로그인

thstkddnr20 opened the pull request:
## 주요 변경 사항
1. JwtAuthenticationFilter에서 Authorization 헤더를 검사하여 토큰을 검증합니다.
2. 로그인 시 JwtTokenProvider에서 토큰을 발급합니다.
3. FilterConfig에 정의된 URL_PATTERNS에만 filter를 적용합니다.
4. 유저에 Role이라는 역할 개념을 추가하였습니다. (ADMIN, USER)
5. 이름에 "카카오"가 들어간 상품 생성, 수정 요청은 관리자가 아닐 경우 불허합니다.

## 맞닥뜨린 문제
1. filter와 interceptor중 filter 선택 이유
jwt 토큰을 검증하되 관심사를 분리하기 위한 방법에는 filter와 interceptor라는 선택지가 있었습니다.
filter와 interceptor에는 DispatcherServlet으로 들어왔냐 안들어왔냐의 차이점이 있습니다.
사용자의 인증 부분을 굳이 DispatcherServlet 내부로 들어와서 할 필요가 없다고 생각하여 filter를 사용하기로 선택하였습니다.

2. filter에서 발생한 예외는 GlobalExceptionHandler에서 처리할 수 없어 ExceptionHandlerFilter로 처리
맞닥뜨린 문제의 1번에서 적은 내용과 이어지는 부분으로, filter는 dispatcherServlet 이전에 실행되기 때문에 RestControllerAdvice를 사용하여 예외를 잡을 수 없는 문제가 있습니다. 이를 해결하기 위해 jwt 검증 필터 앞단에 필터를 하나 더 두고, jwt 필터에서 발생한 예외를 잡아 response를 반환합니다.

3. 테스트 환경에서 @Sql이 정상적으로 실행되지 않음
같은 패키지 안에 있는 테스트 사이에서 1번 테스트 실행 후 2번 테스트 실행할 때 2번 테스트에서 에러가 발생합니다.
이를 해결하기 위해 `@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_CLASS)`를 붙여주어 1개의 테스트 클래스가 끝날경우 ApplicationContext를 재사용하는 것이 아닌 새로 스프링을 띄워 테스트를 진행하였습니다. 이 방식을 적용했더니 모든 테스트가 정상적으로 작동하여 사용 중인데 사실 저 어노테이션을 넣지 않아도 실행되어야 한다고 생각하는데 이 부분에 대해서 확실하게 와닿지 않습니다. 조금 더 찾아보겠습니다.

## 중점적으로 리뷰받고 싶은 곳
1. JwtAuthenticationFilter와 JwtTokenProvider 부분에서 토큰 파싱하는 부분
2. 상품 이름에 카카오가 들어간 생성, 수정 dto 검증에서 적용할 수 있는 코드 중복 방지 방법이 있는지 궁금합니다!
현재의 실행: filter에서 토큰에 이상 없을 시 HttpServletRequest에 setAttribute로 role 집어넣음 -> 상품 이름을 검증해야하는 곳에서 getAttribute로 role을 꺼내옴 -> role 확인 했는데 ADMIN이 아닐 시 상품이름에 "카카오"가 들어갔는지 확인 -> "카카오"가 포함될 경우 에러 처리
3. @DirtiesContext를 이렇게 사용하는게 맞는건지 잘 모르겠습니다 - 서버를 계속 띄우는 것에 있어서 느려진다고 하는데 이렇게 하지 않으면 테스트가 통과하지 않기에 추가해놓은 상태입니다..!

## 추가 내용
secretKey없이 코드 실행이 불가능하므로 secretKey도 올려놓은 상태입니다! -> 이는 원래 원격 저장소에 올리면 안되는 점을 인지하고 있습니다!

Commit:
docs: README.md step2 구현 기능 작성

Commit:
chore: jwt 의존성 추가

Commit:
feat: users 테이블 정의, 유저 회원가입 로직 작성

Commit:
feat: JwtAuthenticationFilter로 사용자 로그인 검증

Commit:
docs: README.md step2 구현 기능 추가 작성

Commit:
fix: 회원가입 path 변경
/api/users -> /api/users/register로 변경

Commit:
fix: DisplayName 잘못된 이름 수정

Commit:
test: UserApiControllerTest e2e 테스트 작성

Commit:
fix: ACCESS_SUBJECT 변수 사용

Commit:
docs: README.md step2 구현 기능 추가 작성

Commit:
feat: 최초 관리자를 추가한 후 역할을 변경할 수 있는 API 생성

Commit:
fix: 이름에 카카오가 들어간 상품은 관리자만 생성 가능하도록 수정
1. 카카오 이름을 검증하는 부분이 적어 코드 중복이 조금 존재하더라도 컨트롤러에서 메서드로 검증
2. 반환 상태코드 변경에 따른 테스트 코드 수정

leaf-upper commented on src/main/resources/application-secret.properties:
```diff
@@ -0,0 +1 @@
+spring.jwt.secret_key=632e3dd54c4d1da3d69ea72a6938e746590b4fafffa8fcff8e6bdd6cd061c2c7834e9ee5cbf6903ff051ccd3e0f588f4f58cabdcf146a31be56b4bab6251f06b
```
좋습니다~ 이런파일은 원격저장소에 올리면 안되는 사실을 잘 알고 계신것 같아요. 주로 환경변수, 마운트 등 여러가지 방식을 통해 서버에 주입하는 방식을 사용해요. 저는 개인적으로 환경변수를 통해 주입하는걸 선호합니다. ㅎㅎ

leaf-upper commented on src/main/java/gift/common/filter/ExceptionHandlerFilter.java:
```diff
@@ -0,0 +1,29 @@
+        }
+    }
+
+    private void setErrorResponse(HttpServletResponse response, Throwable e) throws IOException {
```
좋습니다 저도 개인적으로 RestControllerAdvice를 사용하기보다는 Filter단에서 전역적으로 예외처리하는 방식을 선호해요.
추가적으로 제가 사용하는 방식을 조금더 알려드리면
저는 MappingJackson2JsonView를 통해서 ModelAndView를 만들고
해당 ModelAndView를 HttpServletResponse에 작성하는방식으로 코드를 작성해요
많은 고민을 하신것 같네요 ㅎㅎ

leaf-upper replied:
애플리케이션에서 사용하는 예외클래스를 하나 만들고 해당 예외 클래스를 기반으로 예외처리도 할수있어요 
ex)ApplicationException이라는 클래스를 만들고 애플리케이션 로직의 예외를 해당 예외 기반으로 만듦 -> ApplicationException을 기반으로 예외처리

leaf-upper commented on src/main/java/gift/common/config/AdminUserInitializer.java:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    @Bean
+    CommandLineRunner commandLineRunner() {
```
여러가지 빈 생명주기가 있을것 같아요 CommandLineRunner도 좋지만 ApplicationRunner도 나쁘지 않을것 같네요
추가로 접근제어자 붙여주시면 좋을것 같아요

leaf-upper commented on src/main/java/gift/common/config/FilterConfig.java:
```diff
@@ -0,0 +1,37 @@
+    };
+
+    @Bean
+    public FilterRegistrationBean<Filter> registerExceptionHandlerFilter() {
```
FilterRegistrationBean도 아시는군요 ㅎㅎ
좋은 방식인것 같아요~

leaf-upper commented on src/main/java/gift/common/filter/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,51 @@
+            if (authHeader == null || !authHeader.startsWith(HEADER_PREFIX)) {
+                throw new InvalidAccessTokenException("Authorization 헤더가 유효하지 않습니다.");
+            }
+            final String jwt = authHeader.substring(7);
```
Authorization Header를 정확하게 보내주긴 해야겠네요
7보다는 HEADER_PREFIX.length()가 조금더 괜찮을것 같아요

추가로 trim()과 같은 메서드를 사용해서 좌우공백정도는 제거해줘도 괜찮지 않을까요?

leaf-upper commented on src/main/java/gift/common/filter/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,51 @@
+            }
+            final String jwt = authHeader.substring(7);
+
+            jwtTokenProvider.validAccessToken(jwt);
```
토큰 유효성 검사로직이 있네요 좋습니다.

leaf-upper commented on src/main/java/gift/common/filter/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,51 @@
+            Role role = jwtTokenProvider.getRoleFromToken(jwt);
+            request.setAttribute("role", role);
+
+        } catch (ExpiredJwtException | UnsupportedJwtException | MalformedJwtException | InvalidAccessTokenException | SignatureException e) {
```
해당예외말고 다른 RuntimeException이 발생했을때는 어떻게 처리될까요?
Filter단에서 예외가 발생하는만큼 RestControllerAdvice같은 방식으로는 처리하기 어렵습니다.

leaf-upper commented on src/main/java/gift/controller/ProductApiController.java:
```diff
@@ -23,7 +26,8 @@ public ProductApiController(ProductService productService) {
     @PostMapping
-    public ResponseEntity<ProductResponse> createProduct(@RequestBody @Valid CreateProductRequest request) {
+    public ResponseEntity<ProductResponse> createProduct(@RequestBody @Valid CreateProductRequest request, HttpServletRequest httpServletRequest) {
+        validRoleAndName(httpServletRequest, request.name());
```
상품 이름에 카카오가 들어간 생성, 수정 dto 검증에서 적용할 수 있는 코드 중복 방지 방법이 있는지 궁금합니다! 현재의 실행: filter에서 토큰에 이상 없을 시 HttpServletRequest에 setAttribute로 role 집어넣음 -> 상품 이름을 검증해야하는 곳에서 getAttribute로 role을 꺼내옴 -> role 확인 했는데 ADMIN이 아닐 시 상품이름에 "카카오"가 들어갔는지 확인 -> "카카오"가 포함될 경우 에러 처리

저였다면 어드민이 사용하는 API와 사용자가 사용하는 Controller 및 API 자체를 분리했을것 같아요
어드민이 사용하는 API는 Validation을 하지 않고,
사용자가 사용하는 API는 Name을 검증하는 방식입니다.

Role을 기반으로 사용할 수있는 API를 제한하다면 중복코드는 자연스럽게 제거되지않을까요?

leaf-upper commented on src/test/java/gift/controller/UserApiControllerTest.java:
```diff
@@ -0,0 +1,126 @@
+
+@SpringBootTest
+@AutoConfigureMockMvc
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_CLASS)
```
음 이게 없으면 오류가 나는 이유는 애플리케이션 컨텍스트가 뜨면서 db table을 생성하는데,
테스트가 종료될때 테이블을 제거하면서 다른쪽 테스트에는 테이블이 존재하지 않아서 발생하는것 같아요
dropTable을 하지말고 데이터만 삭제하는방식으로 간다면 이슈가 없지않을까 싶어요

leaf-upper reviewed (changes requested):
안녕하세요 상욱님 Step2고생많으셨습니다.
몇개 코멘트를 남겨드렸는데 확인해주시고 수정요청주시면
재리뷰 하도록할게요! 고생하셨습니다!

Commit:
refactor: 피드백 반영하여 코드 수정
1. CommandLineRunner -> ApplicationRunner 변경
2. 토큰에 trim() 적용
3. 테스트 코드 @Sql dropTable.sql 제거

Commit:
feat: 사용자 인가를 위한 interceptor 도입
AdminOnly라는 어노테이션이 붙은 메서드, 클래스에만 관리자 검사

Commit:
feat: 폼 관리자 화면에도 접근할 수 있도록 관리자 login 구현
jwtFilter에서 쿠키도 검사

thstkddnr20 commented:
![인증인가](https://github.com/user-attachments/assets/6508de3a-5f89-43fe-9cac-7769ccc057cc)
Interceptor로 인가를 검증하는 로직을 그림으로 표현하였습니다.

또한 이전에 구현한 관리자 화면에도 관리자만 접근할 수 있도록 Jwt필터에서 쿠키도 같이 검증하도록 로직을 수정하였습니다.

간단하게 메서드나 클래스에 AdminOnly 어노테이션을 붙이는 것으로 관리자의 인가 검증을 할 수 있습니다.

leaf-upper reviewed (approved):
👍 👍 상욱님 코멘트 주신내용 수정해주신거 확인했습니다.
고생하셨고 다음스텝 진행해주셔도 될것같아요~
승인드리도록 하겠습니다.
